### PR TITLE
Add zero rtt support for client 

### PR DESCRIPTION
**Pull request by lostk1ng**
Motivation:

- Support 0-RTT feature on client side

Modifications:

- Export QUIC session using boringSSL api `SSL_CTX_sess_set_new_cb` and class BoringSSLSessionCallback
- Restore QUIC session using quiche api `quiche_conn_set_session`
- Add EarlyDataSendCallback to send 0-RTT data after send Initial packet
- Add QuicClientSessionCache to store session on client side
- Add 0-RTT example

Result:

- Now it's possible to send 0-RTT packet on client side

### CODE REVIEW

**Comment by normanmaurer** [comment]
@lostk1ng seems like there are test-failures. Can you check ?

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLSessionCallback.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            return;
+        }
+
+        QuicheQuicSslEngine engine = engineMap.get(ssl);
+        if (engine == null) {
+            logger.warn("engine is null ssl: {}", ssl);
+            return;
+        }
+
+        byte[] peerParams = BoringSSL.SSL_get_peer_quic_transport_params(ssl);
```
Comment (normanmaurer) [inline codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLSessionCallback.java:48]: why not just pass these directly to the `newSession(...)` method ? This way we can remove one more JNI call and also reduce the API surface.

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLSessionCallback.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        logger.debug("ssl: {}, session: {}, peerParams: {}", ssl, Arrays.toString(session),
```
Comment (normanmaurer) [inline codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLSessionCallback.java:49]: check `logger.isDebugEnabled()` first to reduce the overhead

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/QuicSslContextBuilder.java [hunk 3]
```diff
@@ -160,6 +160,7 @@ public static QuicSslContext buildForServerWithSni(Mapping<? super String, ? ext
     private String[] applicationProtocols;
     private Boolean earlyData;
     private boolean keylog;
+    private boolean clientSessionCache;
```
Comment (normanmaurer) [inline codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/QuicSslContextBuilder.java:163]: I think we should enable it by default

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/QuicSslContextBuilder.java [hunk 4]
```diff
@@ -179,6 +180,14 @@ public QuicSslContextBuilder earlyData(boolean enabled) {
         return this;
     }
 
+    /**
+     * Enable / disable client session cache.
+     */
+    public QuicSslContextBuilder clientSessionCache(boolean enabled) {
```
Comment (normanmaurer) [inline codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/QuicSslContextBuilder.java:186]: why not use `sessionCache(...)` and allow to enable / disable for client and server ?

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLSessionCallback.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+        byte[] quicSession = toQuicSession(session, peerParams);
+        if (quicSession != null) {
+            logger.info("save session host={}, port={}",
+                    engine.getSession().getPeerHost(), engine.getSession().getPeerPort());
+            sessionCache.saveSession(engine.getSession().getPeerHost(), engine.getSession().getPeerPort(), quicSession);
+        }
+    }
+
+    private static byte[] toQuicSession(byte[] sslSession, byte[] peerParams) {
```
Comment (normanmaurer) [inline codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLSessionCallback.java:60]: Rename to make it clear that how the session is encoded into a `byte[]` is dependent on how quiche does it.

```suggestion
    // Mimic the encoding of quiche: https://github.com/cloudflare/quiche/blob/0.10.0/src/lib.rs#L1668
    private static byte[] toQuicheQuicSession(byte[] sslSession, byte[] peerParams) {
```

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/QuicChannelBootstrap.java [hunk 6]
```diff
@@ -168,6 +169,11 @@ public QuicChannelBootstrap connectionAddress(QuicConnectionAddress connectionAd
         return this;
     }
 
+    public QuicChannelBootstrap earlyDataSendCallBack(EarlyDataSendCallback earlyDataSendCallback) {
```
Comment (normanmaurer) [inline codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/QuicChannelBootstrap.java:172]: nit: javadocs

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/QuicClientSessionCache.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+ * under the License.
+ */
+package io.netty.incubator.codec.quic;
+
+
+import io.netty.util.AsciiString;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+final class QuicClientSessionCache {
```
Comment (normanmaurer) [inline codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/QuicClientSessionCache.java:23]: This cache is currently basically unbound... We should ensure we enforce some bound. Beside this we should also ensure we remove expired sessions after some time. This is to mimic what BoringSSL is doing internally:

https://commondatastorage.googleapis.com/chromium-boringssl-docs/ssl.h.html#SSL_SESS_CACHE_NO_AUTO_CLEAR

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/QuicheQuicSslContext.java [hunk 8]
```diff
@@ -196,7 +199,13 @@ long add(QuicheQuicSslEngine engine) {
      */
     void remove(QuicheQuicSslEngine engine) {
         QuicheQuicSslEngine removed = engineMap.remove(engine.connection.ssl);
-        assert removed == engine;
+        if (removed != null) {
+            assert removed == engine;
+        }
+    }
+
+    public QuicClientSessionCache getSessionCache() {
```
Comment (normanmaurer) [inline codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/QuicheQuicSslContext.java:207]:
```suggestion
   QuicClientSessionCache getSessionCache() {
```

File: codec-native-quic/src/main/c/netty_quic_boringssl.c [hunk 9]
```diff
@@ -597,12 +612,59 @@ void keylog_callback(const SSL* ssl, const char* line) {
...
+    }
+
+    // Execute the java callback
+    (*e)->CallVoidMethod(e, sessionCallback, sessionCallbackMethod, (jlong) ssl, sessionBytes);
+
+    return 0;
+}
+
+jbyteArray netty_boringssl_SSL_get_peer_quic_transport_params(JNIEnv* env, jclass clazz, jlong ssl) {
+    const uint8_t *peer_params = NULL;
```
Comment (normanmaurer) [inline codec-native-quic/src/main/c/netty_quic_boringssl.c:652]: to make it easier to understand why we didn't free the params later include some link to https://commondatastorage.googleapis.com/chromium-boringssl-docs/ssl.h.html#SSL_get_peer_quic_transport_params

File: codec-native-quic/src/main/c/netty_quic_boringssl.c [hunk 10]
```diff
@@ -597,12 +612,59 @@ void keylog_callback(const SSL* ssl, const char* line) {
...
+
+    uint8_t *session_data = NULL;
+    size_t session_data_len = 0;
+    if (SSL_SESSION_to_bytes(session, &session_data, &session_data_len) == 0) {
+        // Get session error
+        return 0;
+    }
+
+    jbyteArray sessionBytes = to_byte_array(e, session_data, session_data_len);
+    OPENSSL_free((void *)session_data);
```
Comment (normanmaurer) [inline codec-native-quic/src/main/c/netty_quic_boringssl.c:639]: Add a link to https://commondatastorage.googleapis.com/chromium-boringssl-docs/ssl.h.html#SSL_SESSION_to_bytes to explain why we need to call free

File: codec-native-quic/src/main/c/netty_quic_boringssl.c [hunk 11]
```diff
@@ -597,12 +612,59 @@ void keylog_callback(const SSL* ssl, const char* line) {
     (*e)->CallVoidMethod(e, keylogCallback, keylogCallbackMethod, (jlong) ssl, keyString);
 }
 
-static jlong netty_boringssl_SSLContext_new0(JNIEnv* env, jclass clazz, jboolean server, jbyteArray alpn_protos, jobject handshakeCompleteCallback, jobject certificateCallback, jobject verifyCallback, jobject servernameCallback, jobject keylogCallback, jint verifyMode, jobjectArray subjectNames) {
+int new_session_callback(SSL *ssl, SSL_SESSION *session) {
```
Comment (normanmaurer) [inline codec-native-quic/src/main/c/netty_quic_boringssl.c:615]: Add a link to https://commondatastorage.googleapis.com/chromium-boringssl-docs/ssl.h.html#SSL_CTX_sess_set_new_cb and explain why we return 0 (as we take no ownership)

File: codec-native-quic/src/test/java/io/netty/incubator/codec/quic/example/QuicClientZeroRTTExample.java [hunk 12]
```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2020 The Netty Project
```
Comment (normanmaurer) [inline codec-native-quic/src/test/java/io/netty/incubator/codec/quic/example/QuicClientZeroRTTExample.java:2]:
```suggestion
 * Copyright 2021 The Netty Project
```

File: codec-native-quic/src/test/java/io/netty/incubator/codec/quic/example/QuicClientZeroRTTExample.java [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+                streamChannel.closeFuture().sync();
+            }
+            quicChannel.closeFuture().sync();
+            channel.close().sync();
+        } finally {
+            group.shutdownGracefully();
+        }
+    }
+
+    static QuicStreamChannel createStream(QuicChannel quicChannel) {
```
Comment (normanmaurer) [inline codec-native-quic/src/test/java/io/netty/incubator/codec/quic/example/QuicClientZeroRTTExample.java:110]:
```suggestion
    private static QuicStreamChannel createStream(QuicChannel quicChannel) {
```

File: codec-native-quic/src/test/java/io/netty/incubator/codec/quic/example/QuicClientZeroRTTExample.java [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+
+    public static void main(String[] args) throws Exception {
+        QuicSslContext context = QuicSslContextBuilder.forClient().trustManager(InsecureTrustManagerFactory.INSTANCE).
+                applicationProtocols("http/0.9").earlyData(true).clientSessionCache(true).build();
+
+        newChannelAndSendData(context, null);
+        newChannelAndSendData(context, new EarlyDataSendCallback() {
+            @Override
+            public void send(QuicChannel quicChannel) {
+                QuicStreamChannel streamChannel = createStream(quicChannel);
```
Comment (normanmaurer) [inline codec-native-quic/src/test/java/io/netty/incubator/codec/quic/example/QuicClientZeroRTTExample.java:48]: calling `createStream()` here is not good as `createStream()` calls `sync()`.

File: codec-native-quic/src/test/java/io/netty/incubator/codec/quic/example/QuicServerZeroRTTExample.java [hunk 15]
```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2020 The Netty Project
```
Comment (normanmaurer) [inline codec-native-quic/src/test/java/io/netty/incubator/codec/quic/example/QuicServerZeroRTTExample.java:2]:
```suggestion
 * Copyright 2021 The Netty Project
```

File: codec-native-quic/src/test/java/io/netty/incubator/codec/quic/example/QuicServerZeroRTTExample.java [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+import java.util.concurrent.TimeUnit;
+
+import io.netty.bootstrap.Bootstrap;
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.*;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.nio.NioDatagramChannel;
+import io.netty.handler.codec.LineBasedFrameDecoder;
+import io.netty.handler.ssl.util.SelfSignedCertificate;
+import io.netty.incubator.codec.quic.*;
```
Comment (normanmaurer) [inline codec-native-quic/src/test/java/io/netty/incubator/codec/quic/example/QuicServerZeroRTTExample.java:28]: don't use wildcard imports

File: codec-native-quic/src/test/java/io/netty/incubator/codec/quic/example/QuicServerZeroRTTExample.java [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+ * under the License.
+ */
+package io.netty.incubator.codec.quic.example;
+
+import java.net.InetSocketAddress;
+import java.util.concurrent.TimeUnit;
+
+import io.netty.bootstrap.Bootstrap;
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.*;
```
Comment (normanmaurer) [inline codec-native-quic/src/test/java/io/netty/incubator/codec/quic/example/QuicServerZeroRTTExample.java:23]: don't use wildcard imports

File: codec-native-quic/src/test/java/io/netty/incubator/codec/quic/example/QuicClientZeroRTTExample.java [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+import io.netty.buffer.Unpooled;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.ChannelInputShutdownReadComplete;
+import io.netty.channel.socket.nio.NioDatagramChannel;
+import io.netty.handler.ssl.util.InsecureTrustManagerFactory;
+import io.netty.incubator.codec.quic.*;
```
Comment (normanmaurer) [inline codec-native-quic/src/test/java/io/netty/incubator/codec/quic/example/QuicClientZeroRTTExample.java:32]: don't use wildcard imports

**Review by normanmaurer — CHANGES_REQUESTED** [review]

**Comment by normanmaurer** [comment]
@lostk1ng great work! There are a few things to fix etc. If you want I can also take it from here and address everything. Whatever you prefer

**Comment by lostk1ng** [comment]
> @lostk1ng great work! There are a few things to fix etc. If you want I can also take it from here and address everything. Whatever you prefer

Sure, please fell free to modify it as I think you can design better api than me for this great feature. Hoping this feature will be merged as soon as possible.

**Comment by normanmaurer** [comment]
@lostk1ng will try to pull it over the finish line this week

**Comment by normanmaurer** [comment]
@lostk1ng PTAL

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/QuicheQuicSslContext.java [hunk 19]
```diff
@@ -95,17 +96,28 @@
...
                 new BoringSSLCertificateVerifyCallback(engineMap, trustManager),
                 mapping == null ? null : new BoringSSLTlsextServernameCallback(engineMap, mapping),
-                keylog ? new BoringSSLKeylogCallback() : null, verifyMode,
+                keylog ? new BoringSSLKeylogCallback() : null,
+                server ? null : new BoringSSLSessionCallback(engineMap, sessionCache), verifyMode,
                 BoringSSL.subjectNames(trustManager.getAcceptedIssuers())));
         apn = new QuicheQuicApplicationProtocolNegotiator(applicationProtocols);
-        this.sessionCacheSize = BoringSSL.SSLContext_setSessionCacheSize(nativeSslContext.address(), sessionCacheSize);
-        this.sessionTimeout = BoringSSL.SSLContext_setSessionCacheTimeout(nativeSslContext.address(), sessionTimeout);
+        if (this.sessionCache != null) {
```
Comment (lostk1ng) [inline codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/QuicheQuicSslContext.java:110]: Should we reuse?
setSessionTimeout((int)sessionTimeout);
setSessionCacheSize((int) sessionCacheSize);

**Comment by lostk1ng** [comment]
@normanmaurer Codes looks good to me overall except one thing to fix.

**Comment by normanmaurer** [comment]
> @normanmaurer Codes looks good to me overall except one thing to fix.

Which one ?

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/QuicheQuicSslEngine.java [hunk 20]
```diff
@@ -283,14 +289,32 @@ private boolean isEmpty(byte[] arr) {
...
-                               byte[][] peerCertificateChain, long creationTime, long timeout) {
+                               byte[][] peerCertificateChain, long creationTime, long timeout, boolean singleUse) {
             synchronized (QuicheQuicSslEngine.this) {
                 initPeerCerts(peerCertificateChain, peerCertificate);
                 this.id = id;
                 this.cipher = cipher;
                 this.protocol = protocol;
                 this.creationTime = creationTime * 1000L;
                 this.timeout = timeout * 1000L;
+                this.invalid = singleUse;
```
Comment (lostk1ng) [inline codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/QuicheQuicSslEngine.java:300]: When I run 0RTT example I can't see 0RTT packet with wireshark, it seems cause by singleUse=true, and so we delete session in cache.

Comment (normanmaurer) [reply codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/QuicheQuicSslEngine.java]: doh! Fixed

**Comment by normanmaurer** [comment]
@lostk1ng fixed the bug... Can you check again ?

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/QuicheQuicSslContext.java (hunk 19 above)
Comment (normanmaurer) [reply codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/QuicheQuicSslContext.java:110]: calling own non static methods in the constructor is not really something I like (due possible NPE etc) so lets keep it

**Comment by lostk1ng** [comment]
> @lostk1ng fixed the bug... Can you check again ?

Code is good, and it works fine!

**Comment by normanmaurer** [comment]
@lostk1ng thanks a lot!

**Pull request merged**